### PR TITLE
Create restart recovery suite as a nightly job 

### PR DESCRIPTION
**Pull request by titusen**
Some tests, which reveal edge cases, take too long for every pr validation, so the nightly job for master is introduced. It runs restart_reovery_test suite.

### CODE REVIEW

File: .github/workflows/restart_recovery_suite.yml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      fail-fast: false
+      matrix:
+            compiler:
+                - "CONCORD_BFT_CONTAINER_CC=clang CONCORD_BFT_CONTAINER_CXX=clang++"
+            ci_build_type:
+                - "-DCMAKE_BUILD_TYPE=RELEASE -DBUILD_COMM_TCP_TLS=FALSE"
+                - "-DCMAKE_BUILD_TYPE=RELEASE -DBUILD_COMM_TCP_TLS=TRUE"
+            use_s3_obj_store:
+                - "-DUSE_S3_OBJECT_STORE=ON"
+                - "-DUSE_S3_OBJECT_STORE=OFF"
```
Comment (teoparvanov) [inline .github/workflows/restart_recovery_suite.yml:33]: Let's simplify the CI test matrix to only test with TLS turned off, and S3 store turned off. 
@titusen and @HristoStaykov . what do you think?

File: .github/workflows/restart_recovery_suite.yml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                              -DENABLE_RESTART_RECOVERY_TESTS=TRUE \
+                              -DBUILD_COMM_TCP_PLAIN=FALSE \
+                              -DCMAKE_CXX_FLAGS_RELEASE='-O3 -g' \
+                              -DUSE_LOG4CPP=TRUE \
+                              -DBUILD_ROCKSDB_STORAGE=TRUE \
+                              ${{ matrix.use_s3_obj_store }} \
+                              -DUSE_OPENTRACING=ON \
+                              -DOMIT_TEST_OUTPUT=OFF\
+                              -DKEEP_APOLLO_LOGS=TRUE\
+                              -DUSE_FAKE_CLOCK_IN_TIME_SERVICE=TRUE\" "\
```
Comment (teoparvanov) [inline .github/workflows/restart_recovery_suite.yml:73]: @titusen and @HristoStaykov , can you please carefully review these parameters together, and make sure each one of them makes sense in the case of the restart recovery suite?

File: .github/workflows/restart_recovery_suite.yml (hunk 1 above)
Comment (HristoStaykov) [reply .github/workflows/restart_recovery_suite.yml:33]: I agree, the tests we currently include focus mostly on corner cases and restart recovery that are independent of the transport layer used. I think also S3 is only used in RO replicas, which we don't have currently in the test suite.

Comment (titusen) [reply .github/workflows/restart_recovery_suite.yml]: Done

File: .github/workflows/restart_recovery_suite.yml (hunk 2 above)
Comment (HristoStaykov) [reply .github/workflows/restart_recovery_suite.yml]: We have double checked that all necessary options are correct. Essentially we build the same way we do in clang's build but activate the restart-recovery suite, which will be the only one that is going to be performed in this run. All looks good.

File: .github/workflows/restart_recovery_suite.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    name: Build
+    timeout-minutes: 600
+    runs-on: ubuntu-18.04
+    outputs:
+      err_output: ${{ steps.prepare_err_artifacts.outputs.test}}
+    strategy:
+      fail-fast: false
+      matrix:
+            compiler:
+                - "CONCORD_BFT_CONTAINER_CC=clang CONCORD_BFT_CONTAINER_CXX=clang++"
```
Comment (teoparvanov) [inline .github/workflows/restart_recovery_suite.yml:27]: Is UDP the default communication layer? Even so, maybe it's worth explicitly stating that the binary is built with the UDP communication module?

File: .github/workflows/restart_recovery_suite.yml [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                              -DBUILD_COMM_TCP_PLAIN=FALSE \
+                              -DCMAKE_CXX_FLAGS_RELEASE='-O3 -g' \
+                              -DUSE_LOG4CPP=TRUE \
+                              -DBUILD_ROCKSDB_STORAGE=TRUE \
+                              -DUSE_S3_OBJECT_STORE=OFF \
+                              -DUSE_OPENTRACING=ON \
+                              -DOMIT_TEST_OUTPUT=OFF\
+                              -DKEEP_APOLLO_LOGS=TRUE\
+                              -DUSE_FAKE_CLOCK_IN_TIME_SERVICE=TRUE\" "\
+              && script -q -e -c "CONCORD_BFT_CTEST_TIMEOUT=18000 make test-single-suite TEST_NAME=skvbc_restart_recovery_tests"
```
[the remark this answers is not in the document]
Comment (titusen) [reply .github/workflows/restart_recovery_suite.yml:69]: I changed all loop counts to 40. Let's see.

File: .github/workflows/restart_recovery_suite.yml (hunk 3 above)
Comment (titusen) [reply .github/workflows/restart_recovery_suite.yml:27]: It's UDP, cause these to flags are set to false.
                              -DBUILD_COMM_TCP_TLS=FALSE \
                              -DBUILD_COMM_TCP_PLAIN=FALSE \

File: tests/apollo/test_skvbc_restart_recovery.py [hunk 5]
```diff
@@ -26,6 +26,9 @@
 
 viewChangeTimeoutSec = 5
 
+loops = 4
+timeouts = 60 # variable accustoms to VM speed
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_restart_recovery.py:30]: Can you explain a bit more how this "accustoms to VM speed" ? Is it an adaptable mechanism, or simply a time buffer?

File: tests/apollo/test_skvbc_restart_recovery.py [hunk 6]
```diff
@@ -58,6 +61,7 @@ def _advance_current_next_primary(new_view, num_replicas):
         _next_primary = (_current_primary + 1) % num_replicas
         return _current_primary, _next_primary
 
+    @unittest.skip("May loop infinitely")
```
[the remark this answers is not in the document]
Comment (titusen) [reply tests/apollo/test_skvbc_restart_recovery.py:64]: On my vm it happened always.

File: tests/apollo/test_skvbc_restart_recovery.py (hunk 5 above)
Comment (titusen) [reply tests/apollo/test_skvbc_restart_recovery.py:30]: It's not adaptable per se. Tt just add this value to vurnable timeouts. U r right that commentis slightly misleading.

File: tests/apollo/test_skvbc_restart_recovery.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (titusen) [reply tests/apollo/test_skvbc_restart_recovery.py:64]: Due to the function being called nursery.start_soon(skvbc.send_indefinite_ops) it is stucked idefinitely. The logic has to be changed for e.g. to use "with_constant_load" decorator. What do u guys think?

**Review by HristoStaykov — APPROVED** [review]
LGTM! Great to see we are almost ready to have this suite run periodically. I have minor recommendations to fix the test that loops indefinitely.

File: tests/apollo/test_skvbc_restart_recovery.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (teoparvanov) [reply tests/apollo/test_skvbc_restart_recovery.py]: @titusen did you address this comment by Hristo in your "final version" commit?

File: tests/apollo/test_skvbc_restart_recovery.py [hunk 7]
```diff
@@ -84,25 +88,31 @@ async def test_restarting_replica_with_client_load(self, bft_network):
...
 
-        for v in range(300):
+        async def client_load(task_status=trio.TASK_STATUS_IGNORED):
+            with trio.CancelScope() as scope:
+                task_status.started(scope)
+                await skvbc.send_indefinite_ops()
+
+        for v in range(loops * 100):
             async with trio.open_nursery() as nursery:
+                scoped_client_load = await nursery.start(client_load)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_restart_recovery.py:98]: Just wondering, why do you prefer to use `start()` versus `start_soon()`, in the context of the nursery?

File: tests/apollo/test_skvbc_restart_recovery.py [hunk 8]
```diff
@@ -84,25 +88,31 @@ async def test_restarting_replica_with_client_load(self, bft_network):
         # uncomment for live tracking of log messages from the test
         # log = foo()
 
-        for v in range(300):
+        async def client_load(task_status=trio.TASK_STATUS_IGNORED):
+            with trio.CancelScope() as scope:
+                task_status.started(scope)
+                await skvbc.send_indefinite_ops()
+
+        for v in range(loops * 100):
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_restart_recovery.py:96]: So with the current parameters, this would loop for 1600 iterations?

Comment (titusen) [reply tests/apollo/test_skvbc_restart_recovery.py:96]: Correct

File: tests/apollo/test_skvbc_restart_recovery.py (hunk 7 above)
Comment (teoparvanov) [inline tests/apollo/test_skvbc_restart_recovery.py:98]: Wouldn't it be cleaner to start the nursery outside of the main loop? If I understand correctly, it's a sort of constant background input rate, which doesn't depend on where you are in the iteration.

File: tests/apollo/test_skvbc_restart_recovery.py (hunk 8 above)
Comment (titusen) [reply tests/apollo/test_skvbc_restart_recovery.py:96]: It is quite fast.

File: tests/apollo/test_skvbc_restart_recovery.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (titusen) [reply tests/apollo/test_skvbc_restart_recovery.py]: Yes

File: tests/apollo/test_skvbc_restart_recovery.py (hunk 8 above)
Comment (teoparvanov) [reply tests/apollo/test_skvbc_restart_recovery.py:96]: OK, thanks for the clarification.

File: tests/apollo/test_skvbc_restart_recovery.py (hunk 7 above)
Comment (titusen) [reply tests/apollo/test_skvbc_restart_recovery.py]: Yep, corrected.

Comment (teoparvanov) [reply tests/apollo/test_skvbc_restart_recovery.py]: Thanks.

File: tests/apollo/test_skvbc_restart_recovery.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (teoparvanov) [reply tests/apollo/test_skvbc_restart_recovery.py]: Thanks.

**Review by teoparvanov — APPROVED** [review]
LGTM, thanks @titusen ! Let's merge, and observe the stability of the nightly job over the next couple of days.
